### ISSUE TITLE
not json-api compliant for single resource endpoints

### ISSUE THREAD

**Issue by mradamh** [botandrose/eboshi_api_shootout#45]
per json-api spec:

> Primary data MUST be either:
> 
> * a single resource object, a single resource identifier object, or null, for requests that target single resources
> * an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections

single resource endpoints (e.g. POST /api/clients) are expecting an array of resources in the tests though.  I can certainly update tests but don't want to break everyone's implementation.

aside - i don't really care either way since this is for fun.  but given README says `The API response must conform to the jsonapi.org standard.`, just pointing it out.

**Comment by botandrose** [comment]
Yes, thank you for investigating and bringing this up. I want to conform with the spec, and I have definitely experienced some uncertainty regarding the handling of singular vs multiple resources.

I _think_ we're currently conforming to json-api, because I interpret that distinction as `/api/client` vs `/api/clients`, not how many resources are returned. For example, to my understanding, the `/api/auth` and `/api/account` resources are singular, while the `/api/clients` resource is plural. I think the `/api/clients/1` resource would be considered singular as well. The lines you've quoted seem to support that in the latter clauses: `for requests that target single resources` etc.

And, specifically regarding `POST /api/clients`, I think we're conforming, because that resource is plural no matter how you're verbing it. We could potentially create multiple clients in one `POST`, so the response should undoubtedly be plural there.

TL;DR: I think single vs plural applies to urls, not responses. What do you think?

**Comment by mradamh** [comment]
All endpoint *urls* are pluralized -- we are acting on a collection of clients in our persistent store.

GET /api/clients - multi (array)
GET /api/clients/{id} - single
POST /api/clients - single
DELETE /api/clients/{id} - single
PUT /api/clients/{id} - single

there is no `api/client`

specifically re: `POST`:
> A resource can be created by sending a POST request to a URL that represents a collection of resources. The request MUST include a single resource object as primary data. The resource object MUST contain at least a type member.

**Comment by botandrose** [comment]
I just read TFA and related issues, and you are definitely correct! Let's change it to conform.

`POST /api/clients` being always (at least potentially) plural makes sense to me, because I think a REST resource is defined by its url, not by its behavior, but it turns out that there be dragons when trying to codify behavior like JSON API is trying to do, see json-api/json-api#202 json-api/json-api#795. Looks like they're trying to add it to version 1.1 of the standard, at least.

**Comment by botandrose** [comment]
Also, just found this. Classified as experimental, though, so YAGNI. http://jsonapi.org/extensions/bulk/

Gonna whip up a PR to change the test and all the implementations.

**Comment by botandrose** [comment]
Reverted 7f3c52d7 and now getting an error from Phoenix. Phoenix thinks I'm an HTML client so its rendering an HTML error page, which makes it nearly impossible to understand what the actual error is during test failure. Can you get it to return "curl-friendly" errors?

**Comment by botandrose** [comment]
Turns out you're the only implementation that has actually implemented this endpoint! Aside from my secret ruby impl that Paul is no longer allowed to copy-paste from. ;) Would you mind putting the PR together to fix this?

**Comment by mradamh** [comment]
> Would you mind putting the PR together to fix this?

sure.  i'll scrape something together tomorrow night.

**Comment by botandrose** [comment]
Cool. Thanks again for raising this issue and shining a clarifying light on json api.
